### ISSUE TITLE
Build piston from scratch - how to install runtimes?

### ISSUE THREAD

**Issue by rizal-rovins** [engineer-man/piston#657]
Hey guys,
I'm trying to build piston from scratch , my purpose is to make some modifications and install additional libraries.

1. I first clone the repo,
2. install the dependencies : cd cli && npm i && cd -
3. Start the build : ./piston start
4. I can see now that my server is running but it has no runtimes.
5. cli/index.js ppman list  > when i run this , it returns empty
6. cli/index.js ppman install python > when i run this , Requested package python-* does not exist

I'm not able to install any packages, what am i missing here?

**Comment by rizal-rovins** [comment]
Okay, I figured it out.
I had to build packages first using :
`./piston build-pkg python 3.9.4`

**Comment by Brikaa** [comment]
You can do `./piston select prod` to run on the production environment and you wouldn't need to build the packages. You would just install them.

**Comment by rhlarora84** [comment]
> You can do `./piston select prod` to run on the production environment and you wouldn't need to build the packages. You would just install them.

Thanks. If I do this, do I need to install packages?

**Comment by Brikaa** [comment]
Yes, you would need to install packages only (not build them) as mentioned [here](https://github.com/engineer-man/piston?tab=readme-ov-file#usage). Packages will be installed from the public packages repo.

**Comment by rhlarora84** [comment]
Thanks Brikaa for your help. That worked. Is there a way to avoid downloading those packages from public packages repo and use the local ones?

**Comment by Brikaa** [comment]
`./piston select dev` to install from the packages you built locally.

**Comment by jackrankin** [comment]
how can you use this to make an image with the builder script? im trying to make an image that has python preinstalled with the custom packages

**Comment by Kaschi14** [comment]
I feel like there is often a terminology issue when talking about packages.
Most people here were probably referring to "python packages" when talking about packages. Whereas Brikaa probably referred to piston packages (the different programming language packages within piston).

If I want to add Python Packages (pip packages) that come preinstalled when using e.g. Python3.12 how can I achieve this?
So far I've tried adding the required packages to build.sh and actually rebuild the image using the `./piston build-pkg` command. But after uninstalling and reinstalling Python3.12 it still didn't have the package in question installed.

My build.sh:
```
#!/bin/bash

PREFIX=$(realpath $(dirname $0))
mkdir -p build

# Install system dependencies for OpenCV
apt-get update
apt-get install -y --no-install-recommends \
    libsm6 \
    libxext6 \
    libxrender-dev \
    libglib2.0-0 \
    libgl1-mesa-glx

cd build
curl "https://www.python.org/ftp/python/3.9.4/Python-3.9.4.tgz" -o python.tar.gz
tar xzf python.tar.gz --strip-components=1
rm python.tar.gz
./configure --prefix "$PREFIX" --with-ensurepip=install
make -j$(nproc)
make install -j$(nproc)
cd ..
rm -rf build 

# Install Python packages
bin/pip3 install numpy scipy pandas pycrypto whoosh bcrypt passlib sympy opencv-python
```

I tried:
```
cli/index.js ppman uninstall python=3.12
cli/index.js ppman install python=3.12
```

And then a request to execute with this body:
```
{
    "language": "python",
    "version": "3.12",
    "files": [
        {
            "name": "my_cool_code.py",
            "content": "import cv2 as cv"
        }
    ]
}
```

So in my case I wanted to add open-cv to the packages. What am I doing wrong?

**Comment by Eli-Ronai** [comment]
[Kaschi14](https://github.com/Kaschi14)

Did you ever find out the answer?

**Comment by SerikaYuzuki** [comment]
Installing python packages, I did first enter the docker container with `docker exec -it piston_api bash` and then use `./pip3 install` in /piston/packages/python/3.*/bin.
I know this is just a work around solution, but at least you can install some python packages.

More solid solution may be using `./piston select dev` and make a custom setting folder in packages then `./piston build-pkg python 3.*` but you will install python from source code and it will cost a day which depends on your pc spec.

I hope this would be a help @Kaschi14

**Comment by Eli-Ronai** [comment]
Thanks!
